lxd/archive/archive.go

```diff
@@ -0,0 +1,254 @@
+package archive
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/lxc/lxd/lxd/apparmor"
+	"github.com/lxc/lxd/lxd/sys"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/ioprogress"
+	"github.com/lxc/lxd/shared/logger"
+	"github.com/lxc/lxd/shared/subprocess"
+)
+
+// ExtractWithFds runs extractor process under specifc AppArmor profile.
+// The allowedCmds argument specify commands which are allowed to run by apparmor.
+// The cmd argument is automatically added to allowedCmds slice.
+func ExtractWithFds(cmd string, args []string, allowedCmds []string, stdin io.ReadCloser, sysOS *sys.OS, output *os.File) error {
+	outputPath := output.Name()
+
+	allowedCmds = append(allowedCmds, cmd)
+	allowedCmdsPath := []string{}
+	for _, c := range allowedCmds {
+		cmdPath, err := exec.LookPath(c)
+		if err != nil {
+			return fmt.Errorf("Failed to start unpack: Failed to find executable: %w", err)
+		}
+		allowedCmdsPath = append(allowedCmdsPath, cmdPath)
+	}
+
+	err := apparmor.ArchiveLoad(sysOS, outputPath, allowedCmdsPath)
+	if err != nil {
+		return fmt.Errorf("Failed to start unpack: Failed to load profile: %w", err)
+	}
+	defer apparmor.ArchiveDelete(sysOS, outputPath)
+	defer apparmor.ArchiveUnload(sysOS, outputPath)
+
+	p, err := subprocess.NewProcessWithFds(cmd, args, stdin, output, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to start unpack: Failed to creating subprocess: %w", err)
+	}
+
+	p.SetApparmor(apparmor.ArchiveProfileName(outputPath))
+
+	err = p.Start()
+	if err != nil {
+		return fmt.Errorf("Failed to start unpack: Failed running: tar: %w", err)
+	}
+
+	p.Wait(context.Background())
+	return nil
+}
+
+// CompressedTarReader returns a tar reader from the supplied (optionally compressed) tarball stream.
+// The unpacker arguments are those returned by DetectCompressionFile().
+// The returned cancelFunc should be called when finished with reader to clean up any resources used. This can be
```

Comment:
Strange wrapping on this line, can "This can be..." be moved onto the next line?

Reply:
Fixed.